_posts/2024-08-06-quarkus-and-leyden.adoc

```diff
@@ -0,0 +1,275 @@
+Essentially, some parts of code might get recompiled multiple times and occasionally revert to a lower tier: it's an highly dynamic process.
+
+Peak optimization is reached when most of the running code is compiled at the highest tier, and background compilation activities become very rare or, ideally, none at all.
+
+Compiling code for peak performance also requires quite some resources, so performing this work ahead of time can also save precious CPU cycles during the application bootstrap, and can manifest in substantial memory savings as well.
+
+But there are some limitations on what we can optimise before runtime just by examining the bytecode. For example, extensive use of reflection prevents the compiler from predicting which symbols will be loaded, linked, and most used at runtime.
+
+The Leyden project has already sucessfully prototyped shifting the work of method compilation from JIT to AOT. Execution and compilation of methods is tracked during the training run. At the end of the run any associated profiling information and compiled code for the method are saved to the AOT archive, allowing them to be quickly mapped back into memory and reused when the application is next run.
+
+As with AOT loading and linking, the training run enables some of the work of profiling and compiling to be done AOT but allows anything not trained still to be compiled via the regular JIT compilation process. Note that method code does not need to have been compiled at the highest tier in order to be saved. Also, when code compiled at a lower tier is restored it can still be recompiled at a higher level.
+
+It can also be deoptimized and re-optimized to adapt to different runtime conditions, just as with code compiled in the current runtime. So, the use of AOT compilation is fully integrated into OpenJDK's adaptive, dynamic compilation and recompilation model: even if some assumptions made during AOT compilation turn out to be suboptimal, the just-in-time compiler can intervene at runtime and improve the code with the new information.
+
+== Current status of Project Leyden
+
+There are already experimental https://jdk.java.net/leyden/[early-access builds of Leyden] that can be tested based on https://openjdk.org/jeps/8315737[this draft JEP about Ahead-of-Time Class Linking]. With the https://www.youtube.com/watch?v=NlJK5BKXtHI[Leyden Project], the training run idea has been extended to a wider range of data structures embedded in the new AOT cache. Now the training data contains:
```

Comment:
This is not a good thing to link to as talks too much about CDS and does not emphasize AOT vs JIT (I am currently discussing this with Sanne and Maria on Slack)

Reply:
Also, 'traingin data' is actually an implementation term that we have not defined. Better to change the itor to the bullets to say

'Now the cache contains the following data:'

n.b. this section delivers on the 'most interestingly' I suggested that we add above to clarify that we focus above on addition of AOT linkage info and compiled code but that there is also more.